runtime: port ctrlport monitor to 3.8 and qt5

bastibl opened the pull request:
I guess this requires much more testing. Would appreciate if someone could give it a try.

marcusmueller commented on gnuradio-runtime/python/gnuradio/ctrlport/GNURadioControlPortClient.py:
```diff
@@ -30,6 +30,8 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+from argparse import ArgumentParser
```
I wish I had a larger number of thumbs to signify my agreement with this change.

marcusmueller commented on gnuradio-runtime/python/gnuradio/ctrlport/gr-ctrlport-monitor:
```diff
@@ -602,53 +592,49 @@ class UpdaterWindow(QtGui.QDialog):
-#             else:
-#                 sys.exit(1)
-#             return
+        except Exception as e:
```
hm, uncommenting this error handling code looks better

bastibl replied:
I'm not sure if I get you right. I think it is uncommented (and in the more recent version, it might even work).

bastibl commented:
I pushed some more changes. In summary:
- ported perf-monitor as well
- move argument parsing from `GNURadioControlPortClient` to the actual application. I think that's better than parsing arguments in a function that is supposed to open the connection.

Since you commented on error handling, I had another look:
- Monitors (like gr-ctrlport-monitor and gr-perf-monitor) are supprocesses, but GRC `kill()`s the flowgraph. I don't think that this allows proper shutdown of child processes, so I switched to `terminate()` and added signal handlers for `SIGINT` and `SIGTERM`. I think the current behavior is: flow graph gets `kill()`ed -> monitor crashed.

The intended  behavior is:
- it's possible to shutdown a QT flow graph by pressing `CTRL-C` in the terminal.
- if you start a flowgraph with  a monitor, the monitor is closed when the flowgraph terminates.
- if you start a flowgraph and connect with a monitor, it just closes the corresponding subwindows but not the monitor (this was the error handler Marcus commented on)
- if you send `SIGTERM` or `SIGINT` with something like the `kill` command the flow graph terminates

mbr0wn commented:
I always get this:
```
Traceback (most recent call last):
  File "/home/mbr0wn/src/gnuradio/gr-blocks/examples/ctrlport/pfb_sync_test_qt.py", line 260, in <module>
    main()
  File "/home/mbr0wn/src/gnuradio/gr-blocks/examples/ctrlport/pfb_sync_test_qt.py", line 255, in main
    tb.blocks_ctrlport_monitor_0.start()
  File "/home/mbr0wn/prefix/default/lib/python3.6/dist-packages/gnuradio/ctrlport/monitor.py", line 44, in start
    print("monitor::endpoints() = %s" % (gr.rpcmanager_get().endpoints()))
AttributeError: 'NoneType' object has no attribute 'endpoints'
ctrlport monitor received shutdown signal
```

Probably unrelated to this PR, but I can't test it.

bastibl commented:
Thanks for giving it a try. Maybe you have CtrlPort but not Thrift? Anyway, I think the `print` should go in the `try-except` block and the error message more verbose. I'll fix it.

mbr0wn commented:
@bastibl I think I have Thrift, but I'll need to double-check my PyBOMBS.

mbr0wn commented:
FYI, I had Thrift, but not Python3-Thrift. So that was my original issue. Now I'm just trying to fight the tools to run the damn thing :(

mormj commented:
I'm getting this printout when running a flowgraph with ctrlport probes and ctrlport monitor:

```
ControlPort Monitor running.
gr::log :INFO: controlport - Apache Thrift: -h josh-p51 -p 33587
monitor::endpoints() = -h josh-p51 -p 33587
Traceback (most recent call last):
  File "comparing_resamplers.py", line 102, in <module>
    main()
  File "comparing_resamplers.py", line 91, in main
    (tb.blocks_ctrlport_monitor_0).start()
  File "/share/gnuradio/grdev/lib/python3.6/dist-packages/gnuradio/ctrlport/monitor.py", line 53, in start
    gr.rpcmanager_get().endpoints())[0]
TypeError: 'map' object is not subscriptable
ctrlport.monitor received shutdown signal
```

tools related as well?

mormj commented:
Once I got my pybombs setup pointing to the right directory, the new ctrlport-monitor appears to be working for me

mormj commented:
Tested the following under Ubuntu 18.04.  From the description of intended behavior above:

- it's possible to shutdown a QT flow graph by pressing CTRL-C in the terminal.
[WORKS]
- if you start a flowgraph with a monitor, the monitor is closed when the flowgraph terminates.
-- from GRC, the flowgraph continues to run in the background when the monitor is closed - but this is likely a different issue not related to the PR 
-- from terminal, the behavior is as expected - monitor and flowgraph are closed
- if you start a flowgraph, connect with a monitor, and close the flowgraph; it just closes the corresponding subwindows but not the monitor (this was the error handler Marcus commented on)
[WORKS]
- if you send SIGTERM or SIGINT with something like the kill command the flow graph terminates and closes its monitors.
[WORKS]

michaelld commented:
@bastibl This PR broke most if not all of the ctrlport tests as well as most if not all of the examples. At issue is the usage of "argv=argv" ... 'argv' was removed from the API here, but not in those tests and examples. Could you please fix those usages of it in another PR? Thx!

bastibl commented:
@michaelld please see https://github.com/gnuradio/gnuradio/pull/2565
